server/src/main/java/org/elasticsearch/index/engine/PrunePostingsMergePolicy.java

```diff
@@ -221,4 +230,56 @@ public BytesRef getPayload() throws IOException {
             return in.getPayload();
         }
     }
+
+    private static Tuple<Bits, Boolean> applyRetentionQuery(CodecReader reader, Query retentionQuery) throws IOException {
```

Comment:
I think we should do this differently:

 1. lets return only Bits and if it's fully deleted return `Bits.MatchNoBits` then you can just do an instanceof check
 2. execute the query (it's likely it doesn't match) and if it doesn't match just return the incoming live-docs?
 3. I consider the set of docs this query matches low so lets build a different Bits instance that uses less memory from it and intersect it with the original bits instead of rebuilding it entirely by wrapping the two sets?

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/43051/commits/b9da89a944f4578033bb7fb0c7e16ee3f1b10b5b.